ABSD substantial adaptation of sbase0.1 tar (MIT; userland/sbase.LICENSE).
Controller source choice explicitly authorizes replacing record validation,
I/O, effect ordering and directory-fd traversal. This is not a smallpatch
or a claim that the original parser/extractor remains unchanged.
Preserves uncompressed ustar create/list/extract and file/directory contents.
Native metadata and stable-pathname policy: userland/README.md, tar section.
No new POSIX API; unsupported compression/extensions/types/options refuse.
Original sources staged by userland/PIN; tests userland/test_tar.py.

--- a/tar.c
+++ b/tar.c
@@ -1,628 +1,272 @@
-/* See LICENSE file for copyright and license details. */
+/* See LICENSE file for copyright and license details.
+ * Substantial ABSD adaptation of sbase 0.1 tar.c (MIT). Native metadata
+ * policy; checked record I/O/validation and path traversal replace the
+ * original mechanisms. Stable pathname bindings required; not atomic.
+ */
 #include <sys/stat.h>
-#include <sys/time.h>
 #include <sys/types.h>
-#ifndef major
-#include <sys/sysmacros.h>
-#endif
-
+#include <dirent.h>
 #include <errno.h>
 #include <fcntl.h>
-#include <grp.h>
-#include <libgen.h>
-#include <pwd.h>
+#include <limits.h>
+#include <stdint.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
 #include <unistd.h>
-
-#include "fs.h"
 #include "util.h"
 
 #define BLKSIZ 512
-
-enum Type {
-	REG       = '0',
-	AREG      = '\0',
-	HARDLINK  = '1',
-	SYMLINK   = '2',
-	CHARDEV   = '3',
-	BLOCKDEV  = '4',
-	DIRECTORY = '5',
-	FIFO      = '6',
-	RESERVED  = '7'
+#define PATHLEN 256 /* ustar name100 + slash + prefix155 */
+struct header {
+ char name[100], mode[8], uid[8], gid[8], size[12], mtime[12], chksum[8];
+ char type, linkname[100], magic[6], version[2], uname[32], gname[32];
+ char major[8], minor[8], prefix[155], padding[12];
 };
-
-struct header {
-	char name[100];
-	char mode[8];
-	char uid[8];
-	char gid[8];
-	char size[12];
-	char mtime[12];
-	char chksum[8];
-	char type;
-	char linkname[100];
-	char magic[6];
-	char version[2];
-	char uname[32];
-	char gname[32];
-	char major[8];
-	char minor[8];
-	char prefix[155];
-};
-
-static struct dirtime {
-	char *name;
-	time_t mtime;
-} *dirtimes;
-
-static size_t dirtimeslen;
-
-static int tarfd;
+_Static_assert(sizeof(struct header) == BLKSIZ, "ustar record");
+static int tarfd, vflag, archive_identity;
 static ino_t tarinode;
 static dev_t tardev;
 
-static int mflag, vflag;
-static int filtermode;
-static const char *filtertool;
-
-static const char *filtertools[] = {
-	['J'] = "xz",
-	['Z'] = "compress",
-	['a'] = "lzma",
-	['j'] = "bzip2",
-	['z'] = "gzip",
-};
-
-static void
-pushdirtime(char *name, time_t mtime)
-{
-	dirtimes = ereallocarray(dirtimes, dirtimeslen + 1, sizeof(*dirtimes));
-	dirtimes[dirtimeslen].name = estrdup(name);
-	dirtimes[dirtimeslen].mtime = mtime;
-	dirtimeslen++;
-}
-
-static struct dirtime *
-popdirtime(void)
-{
-	if (dirtimeslen) {
-		dirtimeslen--;
-		return &dirtimes[dirtimeslen];
-	}
-	return NULL;
-}
-
-static int
-comp(int fd, const char *tool, const char *flags)
-{
-	int fds[2];
-
-	if (pipe(fds) < 0)
-		eprintf("pipe:");
-
-	switch (fork()) {
-	case -1:
-		eprintf("fork:");
-	case 0:
-		dup2(fd, 1);
-		dup2(fds[0], 0);
-		close(fds[0]);
-		close(fds[1]);
-
-		execlp(tool, tool, flags, NULL);
-		weprintf("execlp %s:", tool);
-		_exit(1);
-	}
-	close(fds[0]);
-	return fds[1];
-}
-
-static int
-decomp(int fd, const char *tool, const char *flags)
-{
-	int fds[2];
-
-	if (pipe(fds) < 0)
-		eprintf("pipe:");
-
-	switch (fork()) {
-	case -1:
-		eprintf("fork:");
-	case 0:
-		dup2(fd, 0);
-		dup2(fds[1], 1);
-		close(fds[0]);
-		close(fds[1]);
-
-		execlp(tool, tool, flags, NULL);
-		weprintf("execlp %s:", tool);
-		_exit(1);
-	}
-	close(fds[1]);
-	return fds[0];
-}
-
-static ssize_t
-eread(int fd, void *buf, size_t n)
-{
-	ssize_t r;
-
-again:
-	r = read(fd, buf, n);
-	if (r < 0) {
-		if (errno == EINTR)
-			goto again;
-		eprintf("read:");
-	}
-	return r;
-}
-
-static ssize_t
-ewrite(int fd, const void *buf, size_t n)
-{
-	ssize_t r;
-
-	if ((r = write(fd, buf, n)) != n)
-		eprintf("write:");
-	return r;
-}
-
-static void
-putoctal(char *dst, unsigned num, int size)
-{
-	if (snprintf(dst, size, "%.*o", size - 1, num) >= size)
-		eprintf("snprintf: input number too large\n");
-}
-
-static int
-archive(const char *path)
-{
-	char b[BLKSIZ];
-	struct group *gr;
-	struct header *h;
-	struct passwd *pw;
-	struct stat st;
-	size_t chksum, i;
-	ssize_t l, r;
-	int fd = -1;
-
-	if (lstat(path, &st) < 0) {
-		weprintf("lstat %s:", path);
-		return 0;
-	} else if (st.st_ino == tarinode && st.st_dev == tardev) {
-		weprintf("ignoring %s\n", path);
-		return 0;
-	}
-
-	pw = getpwuid(st.st_uid);
-	gr = getgrgid(st.st_gid);
-
-	h = (struct header *)b;
-	memset(b, 0, sizeof(b));
-
-	if (strlen(path) > 255) {
-		const char *reason = "path exceeds 255 character limit";
-		eprintf("malformed tar archive: %s\n", reason);
-	} else if (strlen(path) >= 100) {
-		size_t prefix_len = 155;
-		const char *last_slash = strrchr(path, '/');
-
-		if (last_slash && last_slash < path + prefix_len) {
-		    prefix_len = last_slash - path + 1;
-		}
-
-		/* strlcpy is fine here - for path ONLY -,
-		 * since we're splitting the path.
-		 * It's not an issue if the prefix can't hold
-		 * the full path — name will take the rest. */
-		strlcpy(h->prefix, path, prefix_len);
-		estrlcpy(h->name, path + prefix_len, sizeof(h->name));
-	} else {
-		estrlcpy(h->name, path,		sizeof(h->name));
-	}
-
-	putoctal(h->mode,    (unsigned)st.st_mode & 0777, sizeof(h->mode));
-	putoctal(h->uid,     (unsigned)st.st_uid,         sizeof(h->uid));
-	putoctal(h->gid,     (unsigned)st.st_gid,         sizeof(h->gid));
-	putoctal(h->size,    0,                           sizeof(h->size));
-	putoctal(h->mtime,   (unsigned)st.st_mtime,       sizeof(h->mtime));
-	memcpy(  h->magic,   "ustar",                     sizeof(h->magic));
-	memcpy(  h->version, "00",                        sizeof(h->version));
-	estrlcpy(h->uname,   pw ? pw->pw_name : "",       sizeof(h->uname));
-	estrlcpy(h->gname,   gr ? gr->gr_name : "",       sizeof(h->gname));
-
-	if (S_ISREG(st.st_mode)) {
-		h->type = REG;
-		putoctal(h->size, (unsigned)st.st_size,  sizeof(h->size));
-		fd = open(path, O_RDONLY);
-		if (fd < 0)
-			eprintf("open %s:", path);
-	} else if (S_ISDIR(st.st_mode)) {
-		h->type = DIRECTORY;
-	} else if (S_ISLNK(st.st_mode)) {
-		h->type = SYMLINK;
-		if ((r = readlink(path, h->linkname, sizeof(h->linkname) - 1)) < 0)
-			eprintf("readlink %s:", path);
-		h->linkname[r] = '\0';
-	} else if (S_ISCHR(st.st_mode) || S_ISBLK(st.st_mode)) {
-		h->type = S_ISCHR(st.st_mode) ? CHARDEV : BLOCKDEV;
-		putoctal(h->major, (unsigned)major(st.st_dev), sizeof(h->major));
-		putoctal(h->minor, (unsigned)minor(st.st_dev), sizeof(h->minor));
-	} else if (S_ISFIFO(st.st_mode)) {
-		h->type = FIFO;
-	}
-
-	memset(h->chksum, ' ', sizeof(h->chksum));
-	for (i = 0, chksum = 0; i < sizeof(*h); i++)
-		chksum += (unsigned char)b[i];
-	putoctal(h->chksum, chksum, sizeof(h->chksum));
-	ewrite(tarfd, b, BLKSIZ);
-
-	if (fd != -1) {
-		while ((l = eread(fd, b, BLKSIZ)) > 0) {
-			if (l < BLKSIZ)
-				memset(b + l, 0, BLKSIZ - l);
-			ewrite(tarfd, b, BLKSIZ);
-		}
-		close(fd);
-	}
-
-	return 0;
-}
-
-static int
-unarchive(char *fname, ssize_t l, char b[BLKSIZ])
-{
-	char lname[101], *tmp, *p;
-	long mode, major, minor, type, mtime, uid, gid;
-	struct header *h = (struct header *)b;
-	int fd = -1;
-	struct timespec times[2];
-
-	if (!mflag && ((mtime = strtol(h->mtime, &p, 8)) < 0 || *p != '\0'))
-		eprintf("strtol %s: invalid number\n", h->mtime);
-	if (remove(fname) < 0 && errno != ENOENT)
-		weprintf("remove %s:", fname);
-
-	tmp = estrdup(fname);
-	mkdirp(dirname(tmp), 0777, 0777);
-	free(tmp);
-
-	switch (h->type) {
-	case REG:
-	case AREG:
-	case RESERVED:
-		if ((mode = strtol(h->mode, &p, 8)) < 0 || *p != '\0')
-			eprintf("strtol %s: invalid number\n", h->mode);
-		fd = open(fname, O_WRONLY | O_TRUNC | O_CREAT, 0600);
-		if (fd < 0)
-			eprintf("open %s:", fname);
-		break;
-	case HARDLINK:
-	case SYMLINK:
-		snprintf(lname, sizeof(lname), "%.*s", (int)sizeof(h->linkname),
-		         h->linkname);
-		if (((h->type == HARDLINK) ? link : symlink)(lname, fname) < 0)
-			eprintf("%s %s -> %s:",
-			        (h->type == HARDLINK) ? "link" : "symlink",
-				fname, lname);
-		break;
-	case DIRECTORY:
-		if ((mode = strtol(h->mode, &p, 8)) < 0 || *p != '\0')
-			eprintf("strtol %s: invalid number\n", h->mode);
-		if (mkdir(fname, (mode_t)mode) < 0 && errno != EEXIST)
-			eprintf("mkdir %s:", fname);
-		pushdirtime(fname, mtime);
-		break;
-	case CHARDEV:
-	case BLOCKDEV:
-		if ((mode = strtol(h->mode, &p, 8)) < 0 || *p != '\0')
-			eprintf("strtol %s: invalid number\n", h->mode);
-		if ((major = strtol(h->major, &p, 8)) < 0 || *p != '\0')
-			eprintf("strtol %s: invalid number\n", h->major);
-		if ((minor = strtol(h->minor, &p, 8)) < 0 || *p != '\0')
-			eprintf("strtol %s: invalid number\n", h->minor);
-		type = (h->type == CHARDEV) ? S_IFCHR : S_IFBLK;
-		if (mknod(fname, type | mode, makedev(major, minor)) < 0)
-			eprintf("mknod %s:", fname);
-		break;
-	case FIFO:
-		if ((mode = strtol(h->mode, &p, 8)) < 0 || *p != '\0')
-			eprintf("strtol %s: invalid number\n", h->mode);
-		if (mknod(fname, S_IFIFO | mode, 0) < 0)
-			eprintf("mknod %s:", fname);
-		break;
-	default:
-		eprintf("unsupported tar-filetype %c\n", h->type);
-	}
-
-	if ((uid = strtol(h->uid, &p, 8)) < 0 || *p != '\0')
-		eprintf("strtol %s: invalid number\n", h->uid);
-	if ((gid = strtol(h->gid, &p, 8)) < 0 || *p != '\0')
-		eprintf("strtol %s: invalid number\n", h->gid);
-
-	if (fd != -1) {
-		for (; l > 0; l -= BLKSIZ)
-			if (eread(tarfd, b, BLKSIZ) > 0)
-				ewrite(fd, b, MIN(l, BLKSIZ));
-		close(fd);
-	}
-
-	if (h->type == HARDLINK)
-		return 0;
-
-	times[0].tv_sec = times[1].tv_sec = mtime;
-	times[0].tv_nsec = times[1].tv_nsec = 0;
-	if (!mflag && utimensat(AT_FDCWD, fname, times, AT_SYMLINK_NOFOLLOW) < 0)
-		weprintf("utimensat %s:", fname);
-	if (h->type == SYMLINK) {
-		if (!getuid() && lchown(fname, uid, gid))
-			weprintf("lchown %s:", fname);
-	} else {
-		if (!getuid() && chown(fname, uid, gid))
-			weprintf("chown %s:", fname);
-		if (chmod(fname, mode) < 0)
-			eprintf("fchmod %s:", fname);
-	}
-
-	return 0;
-}
-
-static void
-skipblk(ssize_t l)
-{
-	char b[BLKSIZ];
-
-	for (; l > 0; l -= BLKSIZ)
-		if (!eread(tarfd, b, BLKSIZ))
-			break;
-}
-
-static int
-print(char *fname, ssize_t l, char b[BLKSIZ])
-{
-	puts(fname);
-	skipblk(l);
-	return 0;
-}
-
-static void
-c(int dirfd, const char *name, struct stat *st, void *data, struct recursor *r)
-{
-	archive(r->path);
-	if (vflag)
-		puts(r->path);
-
-	if (S_ISDIR(st->st_mode))
-		recurse(dirfd, name, NULL, r);
-}
-
-static void
-sanitize(struct header *h)
-{
-	size_t i, j;
-	struct {
-		char  *f;
-		size_t l;
-	} fields[] = {
-		{ h->mode,   sizeof(h->mode)   },
-		{ h->uid,    sizeof(h->uid)    },
-		{ h->gid,    sizeof(h->gid)    },
-		{ h->size,   sizeof(h->size)   },
-		{ h->mtime,  sizeof(h->mtime)  },
-		{ h->chksum, sizeof(h->chksum) },
-		{ h->major,  sizeof(h->major)  },
-		{ h->minor,  sizeof(h->minor)  }
-	};
-
-	/* Numeric fields can be terminated with spaces instead of
-	 * NULs as per the ustar specification.  Patch all of them to
-	 * use NULs so we can perform string operations on them. */
-	for (i = 0; i < LEN(fields); i++){
-		for (j = 0; j < fields[i].l && fields[i].f[j] == ' '; j++);
-		for (; j < fields[i].l; j++)
-			if (fields[i].f[j] == ' ')
-				fields[i].f[j] = '\0';
-	}
-}
-
-static void
-chktar(struct header *h)
-{
-	char tmp[8], *err, *p = (char *)h;
-	const char *reason;
-	long s1, s2, i;
-
-	if (h->prefix[0] == '\0' && h->name[0] == '\0') {
-		reason = "empty filename";
-		goto bad;
-	}
-	if (h->magic[0] && strncmp("ustar", h->magic, 5)) {
-		reason = "not ustar format";
-		goto bad;
-	}
-	memcpy(tmp, h->chksum, sizeof(tmp));
-	for (i = 0; i < sizeof(tmp) && tmp[i] == ' '; i++);
-	for (; i < sizeof(tmp); i++)
-		if (tmp[i] == ' ')
-			tmp[i] = '\0';
-	s1 = strtol(tmp, &err, 8);
-	if (s1 < 0 || *err != '\0') {
-		reason = "invalid checksum";
-		goto bad;
-	}
-	memset(h->chksum, ' ', sizeof(h->chksum));
-	for (i = 0, s2 = 0; i < sizeof(*h); i++)
-		s2 += (unsigned char)p[i];
-	if (s1 != s2) {
-		reason = "incorrect checksum";
-		goto bad;
-	}
-	memcpy(h->chksum, tmp, sizeof(h->chksum));
-	return;
-bad:
-	eprintf("malformed tar archive: %s\n", reason);
-}
-
-static void
-xt(int argc, char *argv[], int mode)
-{
-	char b[BLKSIZ], fname[256 + 1], *p;
-	struct timespec times[2];
-	struct header *h = (struct header *)b;
-	struct dirtime *dirtime;
-	long size;
-	int i, n;
-	int (*fn)(char *, ssize_t, char[BLKSIZ]) = (mode == 'x') ? unarchive : print;
-
-	while (eread(tarfd, b, BLKSIZ) > 0 && (h->name[0] || h->prefix[0])) {
-		chktar(h);
-		sanitize(h), n = 0;
-
-		/* small dance around non-null terminated fields */
-		if (h->prefix[0])
-			n = snprintf(fname, sizeof(fname), "%.*s/",
-			             (int)sizeof(h->prefix), h->prefix);
-		snprintf(fname + n, sizeof(fname) - n, "%.*s",
-		         (int)sizeof(h->name), h->name);
-
-		if ((size = strtol(h->size, &p, 8)) < 0 || *p != '\0')
-			eprintf("strtol %s: invalid number\n", h->size);
-
-		if (argc) {
-			/* only extract the given files */
-			for (i = 0; i < argc; i++)
-				if (!strcmp(argv[i], fname))
-					break;
-			if (i == argc) {
-				skipblk(size);
-				continue;
-			}
-		}
-
-		/* ignore global pax header craziness */
-		if (h->type == 'g' || h->type == 'x') {
-			skipblk(size);
-			continue;
-		}
-
-		fn(fname, size, b);
-		if (vflag && mode != 't')
-			puts(fname);
-	}
-
-	if (mode == 'x' && !mflag) {
-		while ((dirtime = popdirtime())) {
-			times[0].tv_sec = times[1].tv_sec = dirtime->mtime;
-			times[0].tv_nsec = times[1].tv_nsec = 0;
-			if (utimensat(AT_FDCWD, dirtime->name, times, 0) < 0)
-				eprintf("utimensat %s:", fname);
-			free(dirtime->name);
-		}
-		free(dirtimes);
-		dirtimes = NULL;
-	}
-}
-
-static void
-usage(void)
-{
-	eprintf("usage: %s [-C dir] [-J | -Z | -a | -j | -z] -x [-m | -t] "
-	        "[-f file] [file ...]\n"
-	        "       %s [-C dir] [-J | -Z | -a | -j | -z] [-h] -c path ... "
-	        "[-f file]\n", argv0, argv0);
-}
-
-int
-main(int argc, char *argv[])
-{
-	struct recursor r = { .fn = c, .follow = 'P', .flags = DIRFIRST };
-	struct stat st;
-	char *file = NULL, *dir = ".", mode = '\0';
-	int fd;
-
-	ARGBEGIN {
-	case 'x':
-	case 'c':
-	case 't':
-		mode = ARGC();
-		break;
-	case 'C':
-		dir = EARGF(usage());
-		break;
-	case 'f':
-		file = EARGF(usage());
-		break;
-	case 'm':
-		mflag = 1;
-		break;
-	case 'J':
-	case 'Z':
-	case 'a':
-	case 'j':
-	case 'z':
-		filtermode = ARGC();
-		filtertool = filtertools[filtermode];
-		break;
-	case 'h':
-		r.follow = 'L';
-		break;
-	case 'v':
-		vflag = 1;
-		break;
-	default:
-		usage();
-	} ARGEND
-
-	if (!mode)
-		usage();
-	if (mode == 'c')
-		if (!argc)
-			usage();
-
-	switch (mode) {
-	case 'c':
-		tarfd = 1;
-		if (file && *file != '-') {
-			tarfd = open(file, O_WRONLY | O_TRUNC | O_CREAT, 0644);
-			if (tarfd < 0)
-				eprintf("open %s:", file);
-			if (lstat(file, &st) < 0)
-				eprintf("lstat %s:", file);
-			tarinode = st.st_ino;
-			tardev = st.st_dev;
-		}
-
-		if (filtertool)
-			tarfd = comp(tarfd, filtertool, "-cf");
-
-		if (chdir(dir) < 0)
-			eprintf("chdir %s:", dir);
-		for (; *argv; argc--, argv++)
-			recurse(AT_FDCWD, *argv, NULL, &r);
-		break;
-	case 't':
-	case 'x':
-		tarfd = 0;
-		if (file && *file != '-') {
-			tarfd = open(file, O_RDONLY);
-			if (tarfd < 0)
-				eprintf("open %s:", file);
-		}
-
-		if (filtertool) {
-			fd = tarfd;
-			tarfd = decomp(tarfd, filtertool, "-cdf");
-			close(fd);
-		}
-
-		if (chdir(dir) < 0)
-			eprintf("chdir %s:", dir);
-		xt(argc, argv, mode);
-		break;
-	}
-
-	return recurse_status;
-}
+static void *alloc(size_t n) {
+ void *p = malloc(n);
+ if (!p) eprintf("out of memory\n");
+ return p;
+}
+static void checked_close(int fd) {
+ if (close(fd)) eprintf("close:");
+}
+static size_t eread(int fd, void *buf, size_t n) {
+ size_t at = 0;
+ while (at < n) {
+  ssize_t got = read(fd, (char *)buf + at, n-at);
+  if (got < 0) { if (errno == EINTR) continue; eprintf("read:"); }
+  if (!got) break;
+  at += (size_t)got;
+ }
+ return at;
+}
+static void ewrite(int fd, const void *buf, size_t n) {
+ size_t at = 0;
+ while (at < n) {
+  ssize_t got = write(fd, (const char *)buf+at, n-at);
+  if (got < 0) { if (errno == EINTR) continue; eprintf("write:"); }
+  if (!got) eprintf("zero-length write\n");
+  at += (size_t)got;
+ }
+}
+static void record(void *b) {
+ if (eread(tarfd, b, BLKSIZ) != BLKSIZ) eprintf("truncated archive record\n");
+}
+static int zero(const unsigned char *b) {
+ size_t i;
+ for (i=0; i<BLKSIZ; i++) if (b[i]) return 0;
+ return 1;
+}
+static uint64_t octal(const char *p, size_t n) {
+ size_t i=0, digits=0; uint64_t v=0;
+ while (i<n && p[i]==' ') i++;
+ while (i<n && p[i]>='0' && p[i]<='7') {
+  if (v > (UINT64_MAX-(unsigned)(p[i]-'0'))/8) eprintf("numeric overflow\n");
+  v=v*8+(unsigned)(p[i++]-'0'); digits++;
+ }
+ if (!digits || i==n) eprintf("invalid unterminated octal field\n");
+ for (;i<n;i++) if (p[i] && p[i]!=' ') eprintf("invalid octal field\n");
+ return v;
+}
+static void putoctal(char *p, size_t n, uint64_t v) {
+ size_t i=n-1;
+ p[i]=0;
+ while (i) { p[--i]=(char)('0'+(v&7)); v>>=3; }
+ if (v) eprintf("number outside ustar field\n");
+}
+static unsigned checksum(const struct header *h) {
+ const unsigned char *p=(const unsigned char *)h; unsigned v=0; size_t i;
+ for (i=0;i<BLKSIZ;i++) v += i>=148 && i<156 ? ' ' : p[i];
+ return v;
+}
+/* Normalize benign ./ and trailing separators, refuse parent/absolute paths.
+ * Never follows a symlink. The caller must keep pathname bindings stable. */
+static void path_name(const char *in, char out[PATHLEN+1]) {
+ size_t at=0;
+ if (!*in || *in=='/') eprintf("invalid absolute/empty member path\n");
+ while (*in) {
+  const char *end=strchr(in,'/'); size_t n=end ? (size_t)(end-in) : strlen(in);
+  if (n==2 && in[0]=='.' && in[1]=='.') eprintf("parent member path refused\n");
+  if (n && !(n==1 && *in=='.')) {
+   if (at+n+(at!=0)>PATHLEN) eprintf("member path exceeds ustar range\n");
+   if (at) out[at++]='/';
+   memcpy(out+at,in,n); at+=n;
+  }
+  if (!end) break;
+  in=end+1;
+ }
+ if (!at) out[at++]='.';
+ out[at]=0;
+}
+static void header_name(struct header *h, const char *path) {
+ size_t n=strlen(path), i;
+ if (n<=sizeof(h->name)) { memcpy(h->name,path,n); return; }
+ for (i=n;i;i--) if (path[i-1]=='/' && i-1<=sizeof(h->prefix) && n-i<=sizeof(h->name)) {
+  memcpy(h->prefix,path,i-1); memcpy(h->name,path+i,n-i); return;
+ }
+ eprintf("member path has no ustar prefix split\n");
+}
+static void decode_name(const struct header *h, char out[PATHLEN+1]) {
+ char raw[PATHLEN+1]; size_t p=strnlen(h->prefix,sizeof(h->prefix)), n=strnlen(h->name,sizeof(h->name)), at=0;
+ if (!n) eprintf("empty member name\n");
+ if (p) { memcpy(raw,h->prefix,p); at=p; raw[at++]='/'; }
+ memcpy(raw+at,h->name,n); raw[at+n]=0; path_name(raw,out);
+}
+/* Ancestors and destination validated before any current-member mutation. */
+static void check_path(const char *name, int directory) {
+ char p[PATHLEN+1]; struct stat st; size_t i,n=strlen(name);
+ memcpy(p,name,n+1);
+ for (i=0;i<=n;i++) if (!p[i] || p[i]=='/') {
+  char save=p[i]; p[i]=0;
+  if (lstat(p,&st)==0) {
+   if (S_ISLNK(st.st_mode) || ((save || directory) ? !S_ISDIR(st.st_mode) : !S_ISREG(st.st_mode)))
+    eprintf("path kind conflict: %s\n",p);
+   if (!save && archive_identity && st.st_dev==tardev && st.st_ino==tarinode) eprintf("archive would overwrite itself: %s\n",p);
+   if (!save && !directory && !(st.st_mode&S_IWUSR)) eprintf("readonly destination: %s\n",p);
+  } else if (errno!=ENOENT) eprintf("stat %s:",p);
+  p[i]=save;
+ }
+}
+static void parents(const char *name, int directory) {
+ char p[PATHLEN+1]; size_t i,n=strlen(name); struct stat st;
+ memcpy(p,name,n+1);
+ for (i=0;i<=n;i++) if (p[i]=='/' || (!p[i] && directory)) {
+  char save=p[i]; p[i]=0;
+  if (lstat(p,&st)<0) {
+   if (errno!=ENOENT || mkdir(p,0755)) eprintf("mkdir %s:",p);
+  } else if (!S_ISDIR(st.st_mode)) eprintf("directory conflict: %s\n",p);
+  p[i]=save;
+ }
+}
+static void payload(uint64_t size, int out) {
+ unsigned char b[BLKSIZ];
+ while (size) {
+  size_t n=size<BLKSIZ ? (size_t)size : BLKSIZ;
+  record(b); if (out>=0) ewrite(out,b,n); size-=n;
+ }
+}
+static void unarchive(const char *name, uint64_t size, uint64_t mode, int directory) {
+ int fd;
+ check_path(name,directory);
+ parents(name,directory);
+ if (directory) return;
+ fd=open(name,O_WRONLY|O_CREAT|O_TRUNC,0644);
+ if (fd<0) eprintf("open %s:",name);
+ payload(size,fd); checked_close(fd);
+ if (chmod(name,(mode&S_IWUSR)?0644:0444)) eprintf("chmod %s:",name);
+}
+static void archive(const char *path) {
+ struct header h; struct stat st, after; uint64_t left; unsigned char b[BLKSIZ]; int fd=-1;
+ if (lstat(path,&st)) eprintf("stat %s:",path);
+ if (archive_identity && st.st_ino==tarinode && st.st_dev==tardev) return;
+ if (!S_ISREG(st.st_mode) && !S_ISDIR(st.st_mode)) eprintf("unsupported source kind: %s\n",path);
+ memset(&h,0,sizeof(h)); header_name(&h,path);
+ if (st.st_size<0 || st.st_mtime<0) eprintf("negative source metadata\n");
+ putoctal(h.mode,sizeof(h.mode),(uint64_t)(st.st_mode&07777));
+ putoctal(h.uid,sizeof(h.uid),(uint64_t)st.st_uid); putoctal(h.gid,sizeof(h.gid),(uint64_t)st.st_gid);
+ left=S_ISREG(st.st_mode)?(uint64_t)st.st_size:0;
+ putoctal(h.size,sizeof(h.size),left); putoctal(h.mtime,sizeof(h.mtime),(uint64_t)st.st_mtime);
+ putoctal(h.major,sizeof(h.major),0); putoctal(h.minor,sizeof(h.minor),0);
+ memcpy(h.magic,"ustar",6); memcpy(h.version,"00",2); h.type=S_ISDIR(st.st_mode)?'5':'0';
+ putoctal(h.chksum,sizeof(h.chksum),checksum(&h));
+ if (S_ISREG(st.st_mode)) {
+  fd=open(path,O_RDONLY); if (fd<0) eprintf("open %s:",path);
+  if (fstat(fd,&after) || after.st_size!=st.st_size || after.st_dev!=st.st_dev || after.st_ino!=st.st_ino)
+   eprintf("source changed before read: %s\n",path);
+ }
+ ewrite(tarfd,&h,BLKSIZ);
+ while (left) {
+  size_t n=left<BLKSIZ?(size_t)left:BLKSIZ;
+  memset(b,0,sizeof(b)); if (eread(fd,b,n)!=n) eprintf("source shortened: %s\n",path);
+  ewrite(tarfd,b,BLKSIZ); left-=n;
+ }
+ if (fd>=0) {
+  if (eread(fd,b,1) || fstat(fd,&after) || after.st_size!=st.st_size) eprintf("source size changed: %s\n",path);
+  checked_close(fd);
+ }
+ if (vflag && fprintf(stderr,"%s\n",path)<0) eprintf("verbose output:");
+}
+static void walk(const char *path) {
+ struct stat st; DIR *dir; struct dirent *entry;
+ archive(path);
+ if (lstat(path,&st)) eprintf("stat %s:",path);
+ if (!S_ISDIR(st.st_mode)) return;
+ dir=opendir(path); if (!dir) eprintf("opendir %s:",path);
+ for (;;) {
+  char child[PATHLEN+1]; size_t a=!strcmp(path,".")?0:strlen(path), b;
+  errno=0; entry=readdir(dir);
+  if (!entry) { if (errno) eprintf("readdir %s:",path); break; }
+  if (!strcmp(entry->d_name,".") || !strcmp(entry->d_name,"..")) continue;
+  b=strlen(entry->d_name);
+  if (a+(a!=0)+b>PATHLEN) eprintf("source path exceeds ustar range\n");
+  memcpy(child,path,a); if (a) child[a++]='/'; memcpy(child+a,entry->d_name,b+1); walk(child);
+ }
+ if (closedir(dir)) eprintf("closedir:");
+}
+static void xt(int argc, char **argv, int mode) {
+ struct header h; char name[PATHLEN+1]; unsigned char *seen=alloc(argc?(size_t)argc:1); int i;
+ memset(seen,0,argc?(size_t)argc:1);
+ for (;;) {
+  uint64_t size, perm; int selected=!argc, directory;
+  record(&h);
+  if (zero((unsigned char *)&h)) { record(&h); if (!zero((unsigned char *)&h)) eprintf("missing second end record\n"); break; }
+  if (memcmp(h.magic,"ustar",6) || memcmp(h.version,"00",2)) eprintf("unsupported archive format\n");
+  if (octal(h.chksum,sizeof(h.chksum))!=checksum(&h)) eprintf("incorrect header checksum\n");
+  perm=octal(h.mode,sizeof(h.mode)); if (perm>07777) eprintf("invalid mode\n");
+  (void)octal(h.uid,sizeof(h.uid)); (void)octal(h.gid,sizeof(h.gid));
+  (void)octal(h.mtime,sizeof(h.mtime)); size=octal(h.size,sizeof(h.size));
+  directory=h.type=='5';
+  if (!directory && h.type!='0' && h.type!='\0') eprintf("unsupported member type\n");
+  if (h.linkname[0]) eprintf("unexpected link target\n");
+  decode_name(&h,name);
+  if (!strcmp(name,".") && !directory) eprintf("non-directory dot member\n");
+  if (directory) size=0; /* ustar directory size is ignored; no data records */
+  for (i=0;i<argc;i++) {
+   size_t n=strlen(argv[i]);
+   if (!strcmp(argv[i],".") || !strcmp(name,argv[i]) || (!strncmp(name,argv[i],n) && name[n]=='/')) selected=seen[i]=1;
+  }
+  if (!selected) payload(size,-1);
+  else if (mode=='t') { if (puts(name)<0) eprintf("listing output:"); payload(size,-1); }
+  else { unarchive(name,size,perm,directory); if (vflag && fprintf(stderr,"%s\n",name)<0) eprintf("verbose output:"); }
+ }
+ for (i=0;i<argc;i++) if (!seen[i]) eprintf("selected member absent: %s\n",argv[i]);
+ free(seen);
+}
+static void usage(void) {
+ eprintf("usage: tar [-C dir] [-v] [-m] -c|-t|-x [-f archive] [file ...]\n");
+}
+int main(int argc, char **argv) {
+ char *file=NULL,*dir=".",mode=0; int own=0,i; struct stat st; char **names;
+ ARGBEGIN {
+ case 'c': case 't': case 'x': if (mode) usage(); mode=ARGC(); break;
+ case 'C': dir=EARGF(usage()); break;
+ case 'f': file=EARGF(usage()); break;
+ case 'v': vflag=1; break;
+ case 'm': break; /* Native extraction never restores Unix timestamps. */
+ default: usage();
+ } ARGEND
+ if (!mode || (mode=='c' && !argc)) usage();
+ names=alloc((argc?(size_t)argc:1)*sizeof(*names));
+ for (i=0;i<argc;i++) { names[i]=alloc(PATHLEN+1); path_name(argv[i],names[i]); }
+ tarfd=mode=='c'?1:0;
+ if (file && strcmp(file,"-")) {
+  tarfd=open(file,mode=='c'?O_WRONLY|O_CREAT|O_TRUNC:O_RDONLY,0644);
+  if (tarfd<0) eprintf("open archive %s:",file);
+  own=1;
+ }
+ if (!fstat(tarfd,&st)) { archive_identity=1; tarinode=st.st_ino; tardev=st.st_dev; }
+ if (chdir(dir)) eprintf("chdir %s:",dir);
+ if (mode=='c') { unsigned char end[BLKSIZ*2]={0}; for (i=0;i<argc;i++) walk(names[i]); ewrite(tarfd,end,sizeof(end)); }
+ else xt(argc,names,mode);
+ if (fflush(stdout) || ferror(stdout) || fflush(stderr) || ferror(stderr)) eprintf("output error\n");
+ if (own) checked_close(tarfd);
+ for (i=0;i<argc;i++) free(names[i]);
+ free(names);
+ return 0;
+}
